viogpu/viogpu3d/BUILDING.md

```diff
@@ -0,0 +1,13 @@
+# Building viogpu3d driver
+
+This repository contains kernel-mode part of viogpu3d driver, and full-driver will be build only if there is user-mode driver dll's available in directory pointed by environment variable `MESA_PATH`.
+
+## Insturction
+1. Create directory for installed files and set environment variable `MESA_PATH` to path of this directory
+2. Clone [Mesa](https://gitlab.freedesktop.org/mesa/mesa)
+3. Create another directory for building mesa
+4. Inside directory from step 3 run: `meson ..\path_to_mesa_repo --prefix=%MESA_PREFIX% -Dgallium-drivers=virgl -Dgallium-d3d10umd=true`
+5. Build and install mesa binaries: `ninja install`
+6. Go to `viogpu` directory of this repository
+7. Run: `.\build_AllNoSdv.bat`
+8. Compiled drivers will be available in `viogpu\viogpu3d\objfre_win10_amd64\amd64\viogpu3d` 
```

Comment:
Hi @max8rr8,

This instruction is not clear to me. Can you please describe it more with directory examples? What dependencies should be installed and how configured? (Example: https://github.com/virtio-win/kvm-guest-drivers-windows/wiki/Building-the-drivers-using-Windows-11-21H2-EWDK). 
I looked into the [mesa compilation guide](https://docs.mesa3d.org/install.html) and it looks completely different.
```
meson setup builddir/
meson compile -C builddir/
sudo meson install -C builddir/
```

Several questions:
1. What is `%MESA_PREFIX%` and where it defined?
3. Can we use any precompiled MESA for Windows (for example https://fdossena.com/?p=mesa/index.frag)

Reply:
> What is %MESA_PREFIX% and where it defined?

%MESA_PREFIX% environment variable is set during build process, it points to directory where mesa installs its files.

> Can we use any precompiled MESA for Windows

I don't think so, as [pointed out in mesa MR](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/24223#:~:text=Yeah%2C%20this%20is%20the%20Windows%2Dy%20way%20of%20doing%20things%2C%20build%20a%20different%20DLL%20for%20the%20specific%20device%20(or%20software/mapping%20implementation)%20that%20you%20want%20to%20target.%20This%20is%20what%20we%20do%20when%20we%20ship%20GLOn12%20as%20an%20ICD%20on%20Windows.) when building user-mode driver we have to build it with specific mesa flags to build **only** virgl driver (to avoid conflicts)

Reply:
# Building instruction 
NOTE 1: this is valid for now, until changes are not in upstream repositories
NOTE 2: this assumes that all build dependencies(meson, WDK, ninja, etc...) are installed

## Part 1: Virglrenderer
On host machine it is required that patched version of virglrenderer is used.
1. Acquire source code `git clone --branch viogpu_win https://gitlab.freedesktop.org/max8rr8/virglrenderer && cd virglrenderer`
2. Create install directory `mkdir install` and build directory: `mkdir build && cd build`
3. Configure build `meson --prefix=$(pwd)/../install` (we set prefix to install libvirglrenderer not globally but to previously created dir install)
4. Compile and install `ninja install`
Now ensure that qemu loads libvirglrenderer from install directory, this can be done by setting `LD_LIBRARY_PATH` to something like `/some/path/to/starter_dir/virglrenderer/install`

## Part 2: Build mesa
Now inside virtual machine with build tools installed create working directory, then inside it (this assumes use of Powershell):
1. Create mesa prefix dir `mkdir mesa_prefix`  and set env `MESA_PREFIX` to its path: `$env:MESA_PREFIX="$PWD\mesa_prefix"`
2. Get patched mesa source code `git clone --depth 10 --branch viogpu_win https://gitlab.freedesktop.org/max8rr8/mesa` and then cd into it `cd mesa`
3. Create build directory `mkdir build && cd build`
4. Configure build `meson .. --prefix=$env:MESA_PREFIX -Dgallium-drivers=virgl -Dgallium-d3d10umd=true`, build options explained:
  * `--prefix=$env:MESA_PREFIX` set installation path to dir created in step 1
  * `-Dgallium-drivers=virgl` build only virgl driver
  * `-Dgallium-d3d10umd=true` build DirectX 10 user-mode driver (opengl one is build by default)
 5. Build and install (to mesa prefix): `ninja install`
 
 ## Part 3: Build driver
 Now that mesa is build and installed into `%MESA_PREFIX%` viogpu3d will be built (in case `%MESA_PREFIX` is not set viogpu3d inf generation is skipped)
 1. Acquire source code `git clone --branch viogpu_win https://github.com/max8rr8/kvm-guest-drivers-windows` and cd into it `cd kvm-guest-drivers-windows`
 2. Go to viogpu `cd viogpu`
 3. (optional, but very useful) setup test code signning from visual studio
 4. Call build `.\build_AllNoSdv.bat`
 
 ## Part 4: Installation
 Now copy `kvm-guest-drivers-windows\viogpu\viogpu3d\objfre_win10_amd64\amd64\viogpu3d` to target VM and install it.